akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1253,6 +1285,22 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
     case ref  => ref
   }
 
+  // keep track of created SubSinkInlets and SubSourceOutlets to make sure we do not leak them
+  // when this stage completes/fails, not threadsafe only accessed from stream machinery callbacks etc.
+  private var _subInletsAndOutlets: List[AnyRef] = Nil
+
+  private def created(inlet: SubSinkInlet[_]): Unit =
+    _subInletsAndOutlets = inlet :: _subInletsAndOutlets
+
+  private def completedOrFailed(inlet: SubSinkInlet[_]): Unit =
+    _subInletsAndOutlets = _subInletsAndOutlets.filterNot(_ eq inlet)
```

Comment:
Are we really ok with scanning here? E.g. HTTP/2 or big pools will have lots of substreams. Could we use a Set instead?

Reply:
Yeah, I mentioned that in a comment on the whole PR, I optimized only for creating streams thinking that they will mostly be few and close relatively quickly (like concat/merge style things) or stay alive until the end and all be closed at once, but I don't really have any data to back that with. Let's go with Set instead.